Docs: Make an actual LogicMixin spec & explanation (#3975)
ArchipelagoMW/Archipelago: docs/world api.md

is automatically extended by the mixin's members. These members should be prefixed with the name of the implementing
world since the namespace is shared with all other logic mixins.

In general, a LogicMixin class should have at least one mutable variable that is tracking some custom state per player,

> as well as an `init_mixin` and a `copy_mixin` function so that this variable gets initialized and copied correctly when

Reviewer: These are different noun types, so you'd want to make them the same, this way I think is best:
```suggestion
as well `init_mixin` and `copy_mixin` functions so that this variable gets initialized and copied correctly when
```

Author: Uhhh that should still say "as well as" right?

Author: ```suggestion
as well as `init_mixin` and `copy_mixin` functions so that this variable gets initialized and copied correctly when
```

Reviewer: 🤦

Using LogicMixin can slow down your code by a lot if you don't use it intelligently.  
For example, you can make use of the fact that `collect` should only unlock things, and `remove` should only
lock things.
In our example, we have two different functions: `get_newly_unlocked_enemies` and `get_newly_locked_enemies`.  
`get_newly_unlocked_enemies` should only consider enemies that are *not already in the set*
and check whether they were **unlocked**.  
`get_newly_locked_enemies` should only consider enemies that are *already in the set*
and check whether they **became locked**.

There are a multitude of other ways you can optimise LogicMixin. Some games use a `mygame_state_is_stale`
variable that they simply set to True in `collect`, `remove`, and `init_mixin`.  
The calls to the actual recalculating functions are then moved to the start of the relevant access rules like this:

```python
def can_defeat_enemy(state: CollectionState, player: int, enemy: str) -> bool:
    if state.custom_state_is_stale[player]:
        state.defeatable_enemies = recalculate_defeatable_enemies(state)
        state.custom_state_is_stale[player] = False

    return enemy in state.defeatable_enemies
```

> This can help quite significantly because it is possible for 0 local access rules to be called between two calls to

Reviewer: Not really a need for "quite"
```suggestion
This can help significantly because it is possible for 0 local access rules to be called between two calls to
```

lock things.
In our example, we have two different functions: `get_newly_unlocked_enemies` and `get_newly_locked_enemies`.  
`get_newly_unlocked_enemies` should only consider enemies that are *not already in the set*
and check whether they were **unlocked**.  
`get_newly_locked_enemies` should only consider enemies that are *already in the set*
and check whether they **became locked**.

There are a multitude of other ways you can optimise LogicMixin. Some games use a `mygame_state_is_stale`
variable that they simply set to True in `collect`, `remove`, and `init_mixin`.  
The calls to the actual recalculating functions are then moved to the start of the relevant access rules like this:

```python
def can_defeat_enemy(state: CollectionState, player: int, enemy: str) -> bool:
    if state.mygame_state_is_stale[player]:
        state.mygame_defeatable_enemies = recalculate_defeatable_enemies(state)
        state.mygame_state_is_stale[player] = False

    return enemy in state.mygame_defeatable_enemies[player]
```

This can help significantly because it is possible for 0 local access rules to be called between two calls to
`collect`, so recalculating on every `collect` is very slow.

> Only use LogicMixin if necessary. There are often other ways to achieve what it does, like making clever use of
> `state.prog_items`, or using event items, pseudo-regions etc.

Reviewer 2: ```suggestion
Only use LogicMixin if necessary. There are often other ways to achieve what it does, like making clever use of
`state.prog_items`, using event items, pseudo-regions, etc.
```

Author: Hmmm, I guess it's fine. In my mind these are two different "groups" of concepts
1. Accessing state data directly (`state.prog_items`, theoretically also stuff like `state.reachable_regions` or `state.blocked_connections`)
2. Using objects from the "normal" classes to express extra logic ("Pseudo" Regions, (Event) Items / Locations)

Reviewer 2: At the very minimum you need to put the , before etc., or instead of etc. you could write out "or other things" or something like that, I think that would get your way of saying it across better.

Author: I think I'm ok with your change (I already committed it), the idea isn't that important to me to convey

is automatically extended by the mixin's members. These members should be prefixed with the name of the implementing
world since the namespace is shared with all other logic mixins.

LogicMixin is handy when your logic is more complex than one-to-one location-item relationships.  
A game in which "The red key opens the red door" can just express this relationship through a one-line access rule.  
But now, consider a game with a heavy focus on combat, where the main logical consideration is which enemies you can
defeat with your current items.  
There could be dozens of weapons, armor pieces, or consumables that each improve your ability to defeat
specific enemies to varying degrees. It would be useful to be able to keep track of "defeatable enemies" as a state variable,
and have this variable be recalculated as necessary based on newly collected/removed items.

> This is the capability of LogicMixin: Adding custom variables to state that get recalculated "as necessary".

Reviewer 2: ```suggestion
This is the capability of LogicMixin: Adding custom variables to state that get recalculated as necessary.
```
Forgot about the quotes here too, since they don't make sense alone

        }

    def copy_mixin(self, new_state: CollectionState) -> CollectionState:
        # Be careful to make a "deep enough" copy here!
        new_state.mygame_defeatable_enemies = {
            player: enemies.copy() for player, enemies in self.mygame_defeatable_enemies.items()
        }
```

After doing this, you can now access `state.mygame_defeatable_enemies[player]` from your access rules.

Usually, doing this coincides with an override of `World.collect` and `World.remove`, where any time a relevant item is
collected or removed, the custom state variable gets recalculated.

```python
# __init__.py

def collect(self, state: CollectionState, item: Item) -> bool:
    change = super().collect(state, item)
    if change and item in COMBAT_ITEMS:
        state.mygame_defeatable_enemies[self.player] |= get_newly_unlocked_enemies(state)
    return change

def remove(self, state: CollectionState, item: Item) -> bool:
    change = super().remove(state, item)
    if change and item in COMBAT_ITEMS:
        state.mygame_defeatable_enemies[self.player] -= get_newly_locked_enemies(state)
    return change
```

> Using LogicMixin can slow down your code by a lot if you don't use it intelligently. This is because `collect`

Reviewer 2: ```suggestion
Using LogicMixin can slow down your code greatly if you don't use it well. This is because `collect`
```
More misc wording suggestions

Author: I like "greatly" but I think I will keep "intelligently"
